app/processing/optimize/algos.py

```diff
@@ -0,0 +1,352 @@
+                size:
+            ]
+            self.ancilla_edges.append(
+                AncillaConnection(
+                    (source.raw, source_ids),
+                    (self.current_node.raw, target_ids),
+                ),
+            )
+            if len(source.qubit.returned_reusable_after_uncompute_ids) == 0:
+                self.uncomputable.remove(source)
+
+        while len(need_dirty) > 0 and len(self.reusable) > 0:
+            source = self.reusable[0]
+            possible_source_ids = source.qubit.returned_reusable_ids
+            size = min(len(need_dirty), len(possible_source_ids))
+            target_ids = need_dirty[:size]
+            need_dirty = need_dirty[size:]
+            source_ids = possible_source_ids[:size]
+            source.qubit.returned_reusable_ids = possible_source_ids[size:]
+            self.ancilla_edges.append(
+                AncillaConnection(
+                    (source.raw, source_ids),
+                    (self.current_node.raw, target_ids),
+                ),
+            )
+            if len(source.qubit.returned_reusable_ids) == 0:
+                self.reusable.remove(source)
+
+    def satisfy_reusable_qubit_requirement(self) -> None:
+        """Try to satisfy requirement for dirty qubits.
+
+        Add ancilla edges from previous nodes for this case.
+        Try to satisfy with reusable qubits, uncompute nodes if that is not enough.
+        """
+        if self.current_node is None:
+            raise RuntimeError
+
+        need_reusable = self.current_node.qubit.required_reusable_ids
+
+        while len(need_reusable) > 0 and (
+            len(self.uncomputable) > 0 or len(self.reusable) > 0
+        ):
+            while len(need_reusable) > 0 and len(self.reusable) > 0:
+                source = self.reusable[0]
+                possible_source_ids = source.qubit.returned_reusable_ids
+                size = min(len(need_reusable), len(possible_source_ids))
+                target_ids = need_reusable[:size]
+                need_reusable = need_reusable[size:]
+                source_ids = possible_source_ids[:size]
+                source.qubit.returned_reusable_ids = possible_source_ids[size:]
+                self.ancilla_edges.append(
+                    AncillaConnection(
+                        (source.raw, source_ids),
+                        (self.current_node.raw, target_ids),
+                    ),
+                )
+                if len(source.qubit.returned_reusable_ids) == 0:
+                    self.reusable.remove(source)
+
+            if len(need_reusable) > 0 and len(self.uncomputable) > 0:
+                new_reusable = self.pop_uncomputable()
+                self.reusable.append(new_reusable)
+                new_reusable.qubit.returned_reusable_ids.extend(
+                    new_reusable.qubit.returned_reusable_after_uncompute_ids,
+                )
+                new_reusable.qubit.returned_reusable_after_uncompute_ids = []
```

Comment:
Maybe use `clear()`. I think it is more descriptive.
```suggestion
                new_reusable.qubit.returned_reusable_after_uncompute_ids.clear()
```